### DISCUSSION TITLE
mystery email from the depths of the www

### DISCUSSION THREAD

**Discussion by MomsaKoder** [community/community#54421] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

General

### Body

I was trying to add my email to a project( i was learning to do contributions),  but the system managed to create a whole other email from somewhere i cant find.  how do i clear all other emails except my primary email?

after a bit of  poking around its using the first part of my email and the name of my computer..........mind you i have no idea how to fix it

**Comment by spenserblack** [comment]
So I'm assuming you mean the email for the commit's author, right? If not, ignore everything below :laughing:

******

So you've already set your correct email with `git config user.email` (`config --global` to set as the default for all repos) by now, right? If not, make sure you do that :slightly_smiling_face: 

If it's not too many commits, you can do the following:
```bash
git rebase -i --root
# change all actions from `pick` to `edit`

# for each commit
git commit --amend --no-edit --reset-author
git rebase --continue
git push -f # WARNING!
```

To explain these commands:
- `git rebase`
  - Rewrite the git history...
  - `-i`: ...interactively. You write the rebase todo list.
  - `--root`: ...from the first commit.
- The `edit` action will stop at the labelled commit in the rebase todo list and let you edit it.
- `git commit --amend`
  - Change the commit you're on...
  - `--no-edit`: ...but keep the commit message
  - `--reset-author`: ...Set the author's name to the configured user
- `git push -f` (be very careful with this!)
  - Push up your changes. Because you now have a *completely* different history than what was last pushed up, git will not like that, and say you have conflicts. `-f` (or `--force`) basically says "I'm right, the remote is wrong, use my commits."

Why did I add that big `WARNING` comment to the example code? Because it's a destructive action that *overwrites* the remote. If you made a mistake during your rebase (maybe you dropped a commit somewhere) and your local commits are now bad, force-pushing will make your remote's history bad as well, and it can be *very* hard to recover from that. So force-pushing can be risky for beginners, and you should only do it if you *absolutely* know what you're doing. That's not the only reason, and you can simply search "why shouldn't I force push" to find plenty of discussion on the topic. Some will even say to never force push under any circumstance!
So I'd recommend making a branch, or even a new temporary repository, to confirm that you're getting the behavior you want.

So why would you need to force-push? Why would there be a conflict if your changes and commit messages are the same? Commits are identified by their SHAs. The SHAs take a variety of values as seeds to attempt build a unique value (very low chance for duplicates), including
- Time of commit -- once you amend a commit, that's a new time
- Commit message
- And, yes, the author info
sources: [1](https://gist.github.com/masak/2415865), [2](https://stackoverflow.com/a/68806436/19109685)

So git will think the amended commits are completely different commits. In fact, if you ran a `git pull` and fixed the merge conflict with the remote, you would have duplicate commits, one with the old author and one with the new author, everywhere!

That was probably too much info :laughing: but I wanted to really hammer in the unsafe parts and why you should be careful.

The solution I provided (interactive rebase) can be a repetitive, but I wanted to provide something that's step-by-step and not a simple copy+paste job so you have the opportunity to see the changes happen in action, and also quit if you need to with `git rebase --abort`.

[This stackoverflow answer](https://stackoverflow.com/a/1320317/19109685) provides an alternative way to do this which is much less repetitive.

**Comment by spenserblack** [comment]
OK, after digging through your repos I think your problem is *much* simpler than I thought. [Is this the repo in question](https://github.com/MomsaKoder/start-here-guidelines)?
